managed-ledger/src/main/java/org/apache/bookkeeper/mledger/dlog/DlogBasedEntryCacheImpl.java

```diff
@@ -0,0 +1,360 @@
+            log.debug("[{}] Reading entries range : {} to {} in asyncReadEntries of cache", ml.getName(), firstEntry, lastEntry);
+        }
+        Collection<DlogBasedEntry> cachedEntries = entries.getRange(firstPosition, lastPosition);
+
+        if (cachedEntries.size() == entriesToRead) {
+            long totalCachedSize = 0;
+            final List<DlogBasedEntry> entriesToReturn = Lists.newArrayListWithExpectedSize(entriesToRead);
+
+            // All entries found in cache
+            for (DlogBasedEntry entry : cachedEntries) {
+                entriesToReturn.add(DlogBasedEntry.create(entry));
+                totalCachedSize += entry.getLength();
+                entry.release();
+            }
+
+            manager.mlFactoryMBean.recordCacheHits(entriesToReturn.size(), totalCachedSize);
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Entries : {}-{} Found in cache", ml.getName(), firstEntry, lastEntry);
+            }
+            callback.readEntriesComplete((List) entriesToReturn, ctx);
+
+        } else {
+            if (!cachedEntries.isEmpty()) {
+                cachedEntries.forEach(entry -> entry.release());
+            }
+
+            try{
+
+                AsyncLogReader logReader = FutureUtils.result(distributedLogManager.openAsyncLogReader(new DLSN(logSegNo, firstEntry, 0)));
+                //todo do I use futureListener here ok?
+                // Read all the entries from dlog
+                logReader.readBulk(entriesToRead, 100, TimeUnit.MILLISECONDS).whenComplete(new FutureEventListener<List<LogRecordWithDLSN>>() {
+                    @Override
+                    public void onSuccess(List<LogRecordWithDLSN> logRecordWithDLSNs) {
+
+                        checkNotNull(ml.getName());
+                        checkNotNull(ml.getExecutor());
+                        ml.getExecutor().submitOrdered(ml.getName(), safeRun(() -> {
+                            // We got the entries, we need to transform them to a List<> type
+                            final List<DlogBasedEntry> entriesToReturn = Lists.newArrayListWithExpectedSize(entriesToRead);
+                            long totalSize = 0;
+                            Iterator iterator = logRecordWithDLSNs.iterator();
+                            while (iterator.hasNext()){
+                                DlogBasedEntry entry = DlogBasedEntry.create((LogRecordWithDLSN) iterator.next());
+                                entriesToReturn.add(entry);
+                                totalSize += entry.getLength();
+                            }
+//                        update totalSize failure in lambda
+//                        logRecordWithDLSNs.forEach(logRecordWithDLSN -> {
+//                            DlogBasedEntry entry = DlogBasedEntry.create(logRecordWithDLSN);
+//
+//                            entriesToReturn.add(entry);
+//
+//                            totalSize += entry.getLength();
```

Comment:
you can mutate a variable inside a closure. you need to change totalsize to a long holder. e.g. AtomicLong or MutableLong.